File: packages/kari-core/src/kari_core/core/auto_logo.py
"""Shared auto-Logo brand matcher.

Both the V3 renderer (`processor/v3_renderer.py::_resolve_auto_logo_path`) and the
legacy V2 Jinja helper (`core/jinja2renders.py::resolve_auto_logo`) call the same
pure primitive here so that a rename or matcher tweak can never desynchronise the
two Python paths again. Front-end previews mirror the same behaviour through a
shared JSON contract fixture (`tests/fixtures/v3_auto_logo_cases.json`).

V3 policy — implemented at the callsite:
  - `_resolve_auto_logo_path` uses ONLY the builtin logos directory (no `custom/`
    lookup) and never falls back to fujifilm. Unknown or missing Make → None,
    so the preview draws a neutral skeleton and the exported watermark is
    transparent in the same place.
"""
from __future__ import annotations

from collections.abc import Iterable


def _tokenize(source: str) -> set[str]:
    """Return the set of tokens (length > 2) from a free-form brand string.

    Matches the front-end `tokenize` in `apps/web/src/autoLogo.ts` byte-for-byte
    so the two implementations agree on every input.
    """
    return {
        token
        for token in source.lower().replace("-", " ").replace("_", " ").split()
        if len(token) > 2
    }


def _stem_matches_tokens(stem: str, make_tokens: set[str]) -> bool:
    stem_tokens = _tokenize(stem)
    return any(token in stem_tokens for token in make_tokens)


def match_brand_stem(make: str | None, stems: Iterable[str]) -> str | None:
    """Return the first stem whose tokens overlap `make`'s tokens.

    Iteration order is case-insensitive-alphabetical so that changes in
    filesystem enumeration order don't produce different logos for the same
    Make. Returns None when `make` is empty, when no tokens are meaningful, or
    when nothing matches — callers layer their own fallback if their product
    policy requires one (V2 does; V3 does not).
    """
    if not make:
        return None
    tokens = _tokenize(make)
    if not tokens:
        return None
    ordered = sorted(stems, key=str.lower)
    for stem in ordered:
        if _stem_matches_tokens(stem, tokens):
            return stem
    return None

File: apps/web/src/autoLogo.ts
/**
 * V3 auto-Logo resolver — deterministic mirror of the Python V3 policy in
 * `packages/kari-core/src/kari_core/processor/v3_renderer.py::_resolve_auto_logo_path`.
 *
 * V3 policy:
 *   - builtin registry only (mirrors the `/api/logos` list on the front end
 *     and `LOGOS_DIR/*.png,*.jpg,*.jpeg,*.webp` on the back end)
 *   - whole-token, case-insensitive match against builtin stems
 *   - NO fujifilm fallback — unknown/missing Make returns null so the preview
 *     draws a neutral skeleton and PIL Process produces a transparent slot,
 *     matching what the user will actually get.
 *
 * The shared JSON contract lives at
 * `packages/kari-core/tests/fixtures/v3_auto_logo_cases.json`; the Python
 * unit test `test_v3_auto_logo.py` and this file's contract test both iterate
 * that same fixture so a drift on either side breaks CI.
 */

function tokenize(source: string): Set<string> {
  return new Set(
    source
      .toLowerCase()
      .replace(/[-_]/g, ' ')
      .split(/\s+/)
      .filter((token) => token.length > 2),
  );
}

function stemMatchesTokens(stem: string, tokens: Set<string>): boolean {
  const stemTokens = tokenize(stem);
  for (const token of tokens) {
    if (stemTokens.has(token)) return true;
  }
  return false;
}

export type AutoLogoReason = 'no-registry' | 'no-make' | 'matched' | 'no-match';

export interface AutoLogoResolution {
  /** `builtin:<stem>` reference for a matched brand, else null (skeleton). */
  path: string | null;
  reason: AutoLogoReason;
}

/**
 * Resolve the preview-side auto-Logo path for the given EXIF Make and the set
 * of `/api/logos` builtin names.
 *
 * Return values are deliberately narrow — a null path is the correct product
 * outcome for anything the backend would also fail to resolve, so preview and
 * Process never diverge in what the user sees. The `reason` field is for
 * developer tooling / tests, not user-facing copy.
 */
export function resolveAutoLogo(
  make: string | undefined | null,
  builtins: readonly string[],
): AutoLogoResolution {
  if (!builtins || builtins.length === 0) {
    return { path: null, reason: 'no-registry' };
  }
  const trimmed = (make ?? '').trim();
  if (!trimmed) {
    return { path: null, reason: 'no-make' };
  }
  const tokens = tokenize(trimmed);
  if (tokens.size === 0) {
    return { path: null, reason: 'no-make' };
  }
  // Case-insensitive alphabetical order matches Python's sorted(..., key=str.lower).
  const ordered = [...builtins].sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
  for (const stem of ordered) {
    if (stemMatchesTokens(stem, tokens)) {
      return { path: `builtin:${stem}`, reason: 'matched' };
    }
  }
  return { path: null, reason: 'no-match' };
}
